utils.py

```diff
@@ -28,13 +29,14 @@ def preprocess(self, input_file, vocab_file, tensor_file):
         with codecs.open(input_file, "r", encoding=self.encoding) as f:
             data = f.read()
         counter = collections.Counter(data)
+        counter.update(('<S>', '</S>', 'UNK'))  # add tokens for start end and unk
         count_pairs = sorted(counter.items(), key=lambda x: -x[1])
         self.chars, _ = zip(*count_pairs)
         self.vocab_size = len(self.chars)
         self.vocab = dict(zip(self.chars, range(len(self.chars))))
         with open(vocab_file, 'wb') as f:
             cPickle.dump(self.chars, f)
-        self.tensor = np.array(list(map(self.vocab.get, data)))
+        self.tensor = np.array(list(map(self.vocab.get, ['<S>'] + list(data) + ['</S>'])))
```

Comment:
Do you think it would be a better idea to write this after line 59, `self.tensor = self.tensor[:self.num_batches * self.batch_size * self.seq_length]` ?